timestamp_test.go

```diff
@@ -461,7 +461,104 @@ func TestCreateResponseWithNoTSACertificate(t *testing.T) {
 		t.Error("TSA certificate must not be included in timestamp response")
 	}
 }
+func TestCreateResponseIncludeTSAAndIntermediatesCertificate(t *testing.T) {
+	tsakey := getTSARSAChildKey()
+	tsaCert := getTSAChildCert()
+	parentCert := getTSACert()
 
+	h := sha256.New()
+	_, err := h.Write([]byte("Hello World"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	genTime := time.Now().UTC()
+
+	nonce := big.NewInt(0)
+	nonce = nonce.SetBytes([]byte{0x1, 0x2, 0x3})
+
+	duration, _ := time.ParseDuration("1s")
+
+	timestamp := Timestamp{
+		HashAlgorithm:     crypto.SHA256,
+		HashedMessage:     h.Sum(nil),
+		Time:              genTime,
+		Nonce:             nonce,
+		Policy:            asn1.ObjectIdentifier{2, 4, 5, 6},
+		Ordering:          true,
+		Accuracy:          duration,
+		Qualified:         true,
+		AddTSACertificate: true,
+		Certificates:      []*x509.Certificate{parentCert}, // add parent certificate
+	}
+	timestampBytes, err := timestamp.CreateResponse(tsaCert, tsakey)
+	if err != nil {
+		t.Fatalf("unable to generate time stamp response: %s", err.Error())
+	}
+
+	timestampRes, err := ParseResponse(timestampBytes)
+	if err != nil {
+		t.Fatalf("unable to parse time stamp response: %s", err.Error())
+	}
+
+	if timestampRes.HashAlgorithm.HashFunc() != crypto.SHA256 {
+		t.Errorf("expected hash algorithm is SHA256")
+	}
+	if len(timestampRes.HashedMessage) != 32 {
+		t.Errorf("got %d: expected: %d", len(timestampRes.HashedMessage), 32)
+	}
+
+	if timestampRes.Accuracy != duration {
+		t.Errorf("got accuracy %s: expected: %s", timestampRes.Accuracy, duration)
+	}
+
+	if !timestampRes.Qualified {
+		t.Errorf("got %t: expected: %t", timestampRes.Qualified, true)
+	}
+
+	if !timestampRes.AddTSACertificate {
+		t.Error("TSA certificate must be included in timestamp response")
+	}
+	if fmt.Sprintf("%X", timestamp.Certificates[0].SubjectKeyId) != fmt.Sprintf("%X", parentCert.SubjectKeyId) {
+		t.Errorf("parent certificate is not match with %X", parentCert.SubjectKeyId)
+	}
+}
+func getTSARSAChildKey() *rsa.PrivateKey {
```

Comment:
I would be great if we can use more consistent naming here. The test case it called, TestCreateResponseInclude**TSA**And**Intermediates**Certificate but in this code, it's called child and parent keys.

You can use TSA (for the CA) and TSU (for the signing key) as defined in the RFC or you can use the naming as in a hierarchy [root]>[intermediate]>[end-entity] certificate.